access/pagerduty/app.go

```diff
@@ -55,337 +69,537 @@ func (a *App) WaitReady(ctx context.Context) (bool, error) {
+		// If plugin data is empty or missing incidentID, we cannot do anything.
+		if existing == nil {
+			return PluginData{}, false
+		}
+		if incidentID = existing.IncidentID; incidentID == "" {
+			return PluginData{}, false
+		}
 
-	user, err := a.bot.GetUserByEmail(ctx, req.User)
-	if err != nil {
-		if trace.IsNotFound(err) {
-			log.WithError(err).Debugf("Failed to auto-approve the request")
-			return nil
+		// If resolution field is not empty then we already resolved the incident before. In this case we just quit.
+		if existing.RequestData.Resolution.Tag != Unresolved {
+			return PluginData{}, false
 		}
-		return err
-	}
 
-	ctx, log = logger.WithFields(ctx, logger.Fields{
-		"pd_user_email": user.Email,
-		"pd_user_name":  user.Name,
+		// Mark incident as resolved.
+		pluginData := *existing
+		pluginData.Resolution = resolution
+		return pluginData, true
 	})
-
-	isOnCall, err := a.bot.IsUserOnCall(ctx, user.ID)
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	if isOnCall {
-		log.Infof("User is now on-call, auto-approving the request")
-		return a.setRequestState(ctx, req.ID, incidentID, user.Email, access.StateApproved)
+	if !ok {
+		logger.Get(ctx).Debug("Failed to resolve the incident: plugin data is missing")
+		return nil
+	}
+
+	ctx, log := logger.WithField(ctx, "pd_incident_id", incidentID)
+	if err := a.pagerduty.ResolveIncident(ctx, incidentID, resolution); err != nil {
+		return trace.Wrap(err)
 	}
+	log.Info("Successfully resolved the incident")
 
-	log.Debug("User is not on call")
 	return nil
 }
 
-func (a *App) getPluginData(ctx context.Context, reqID string) (PluginData, error) {
-	dataMap, err := a.accessClient.GetPluginData(ctx, reqID)
+// modifyPluginData performs a compare-and-swap update of access request's plugin data.
+func (a *App) modifyPluginData(ctx context.Context, reqID string, fn func(data *PluginData) (PluginData, bool)) (bool, error) {
+	var lastErr error
+	for i := 0; i < maxModifyPluginDataTries; i++ {
+		oldData, err := a.getPluginData(ctx, reqID)
+		if err != nil {
+			return false, trace.Wrap(err)
+		}
+		newData, ok := fn(oldData)
+		if !ok {
+			return false, nil
+		}
+		var expectData PluginData
+		if oldData != nil {
+			expectData = *oldData
+		}
+		err = trace.Wrap(a.updatePluginData(ctx, reqID, newData, expectData))
+		if err == nil {
+			return true, nil
+		}
+		if trace.IsCompareFailed(err) {
+			lastErr = err
+			continue
+		}
+		return false, err
+	}
+	return false, lastErr
+}
+
+// getPluginData loads a plugin data for a given access request. It returns nil if it's not found.
+func (a *App) getPluginData(ctx context.Context, reqID string) (*PluginData, error) {
+	dataMaps, err := a.apiClient.GetPluginData(ctx, types.PluginDataFilter{
+		Kind:     types.KindAccessRequest,
+		Resource: reqID,
+		Plugin:   pluginName,
+	})
 	if err != nil {
-		return PluginData{}, trace.Wrap(err)
+		return nil, trace.Wrap(err)
+	}
+	if len(dataMaps) == 0 {
+		return nil, nil
```

Comment:
Nit: Typically returning (nil, nil) is not advised, I would consider returning trace.NotFound error here and below.